core/lib/libmbedtls/src/tee_lmd_provider.c

```diff
@@ -9,34 +9,195 @@
+	case TEE_ALG_SHA1:
+	case TEE_ALG_DSA_SHA1:
+	case TEE_ALG_HMAC_SHA1:
+		return mbedtls_md_info_from_string("SHA1");
+#endif
+#if defined(CFG_CRYPTO_MD5)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5:
+	case TEE_ALG_MD5:
+	case TEE_ALG_HMAC_MD5:
+		return mbedtls_md_info_from_string("MD5");
+#endif
+#if defined(CFG_CRYPTO_SHA224)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA224:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA224:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA224:
+	case TEE_ALG_SHA224:
+	case TEE_ALG_DSA_SHA224:
+	case TEE_ALG_HMAC_SHA224:
+		return mbedtls_md_info_from_string("SHA224");
+#endif
+#if defined(CFG_CRYPTO_SHA256)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA256:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA256:
+	case TEE_ALG_SHA256:
+	case TEE_ALG_DSA_SHA256:
+	case TEE_ALG_HMAC_SHA256:
+		return mbedtls_md_info_from_string("SHA256");
+#endif
+#if defined(CFG_CRYPTO_SHA384)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA384:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA384:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA384:
+	case TEE_ALG_SHA384:
+	case TEE_ALG_HMAC_SHA384:
+		return mbedtls_md_info_from_string("SHA384");
+#endif
+#if defined(CFG_CRYPTO_SHA512)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA512:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA512:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA512:
+	case TEE_ALG_SHA512:
+	case TEE_ALG_HMAC_SHA512:
+		return mbedtls_md_info_from_string("SHA512");
+#endif
+	case TEE_ALG_RSAES_PKCS1_V1_5:
+		/* invalid one. but it should not be used anyway */
+		return NULL;
+
+	default:
+		return NULL;
+	}
+}
+#endif /*
+	* defined(_CFG_CRYPTO_WITH_HASH) ||
+	* defined(CFG_CRYPTO_RSA) || defined(_CFG_CRYPTO_WITH_MAC)
+	*/
 
 /******************************************************************************
  * Message digest functions
  ******************************************************************************/
 
 #if defined(_CFG_CRYPTO_WITH_HASH)
+
 TEE_Result crypto_hash_get_ctx_size(uint32_t algo, size_t *size)
 {
+	switch (algo) {
+#if defined(CFG_CRYPTO_MD5)
+	case TEE_ALG_MD5:
+#endif
+#if defined(CFG_CRYPTO_SHA1)
+	case TEE_ALG_SHA1:
+#endif
+#if defined(CFG_CRYPTO_SHA224)
+	case TEE_ALG_SHA224:
+#endif
+#if defined(CFG_CRYPTO_SHA256)
+	case TEE_ALG_SHA256:
+#endif
+#if defined(CFG_CRYPTO_SHA384)
+	case TEE_ALG_SHA384:
+#endif
+#if defined(CFG_CRYPTO_SHA512)
+	case TEE_ALG_SHA512:
+#endif
+		*size = sizeof(mbedtls_md_context_t);
+		break;
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
 
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	return TEE_SUCCESS;
 }
 
 TEE_Result crypto_hash_init(void *ctx, uint32_t algo)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int lmd_res;
+	const mbedtls_md_info_t *md_info = NULL;
+
+	if (ctx == NULL)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	md_info = tee_algo_to_mbedtls_hash_info(algo);
+	if (md_info == NULL)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	mbedtls_md_init(ctx);
+
+	lmd_res = mbedtls_md_setup(ctx, md_info, 0);
```

Comment:
Why are we bothering with this extra layer when the `mbedtls_sha{1,256,512}_*` functions and context types already are available? It comes with a cost to use this layer.